### PR TITLE
Fix surface creation bugs in SHPInterface

### PR DESCRIPTION
**Pull request by TomFischer**
- First bug was that empty polygons were created and the AABB of an empty point set is not allowed.
- Second bug was inproper handling parallel line segments.

### CODE REVIEW

**Comment by TomFischer** [comment]
Jenkins, test this please.

**Comment by endJunction** [comment]
:fast_forward:

**Comment by rinkk** [comment]
I have no idea what you're doing but it look okay and it works.

File: GeoLib/AnalyticalGeometry.cpp [hunk 1]
```diff
@@ -133,11 +133,16 @@ bool lineSegmentIntersect(
 	if (parallel(v,w)) {
 		if (parallel(pq,v)) {
 			// check if c is located at v (c-a = t (b-a), t in [0,1])
-			if (qp[0] / v[0] <= 1.0)
+			if (0 <= qp[0] / v[0] && qp[0] / v[0] <= 1.0) {
```
Comment (chleh) [inline GeoLib/AnalyticalGeometry.cpp]: Can `v[0]` be 0.0?
Btw: `0 <= qp[0] / v[0]` essentially checks if ` qp[0] ` and `v[0]` have the same sign. Is that intended?

File: GeoLib/Polygon.cpp [hunk 2]
```diff
@@ -368,43 +368,34 @@ void Polygon::splitPolygonAtIntersection (std::list<Polygon*>::iterator polygon_
...
+			polyline0.addPoint (intersection_pnt_id);
 			for (std::size_t k(idx1 + 1); k < (*polygon_it)->getNumberOfPoints(); k++)
-				polygon0->addPoint ((*polygon_it)->getPointID (k));
-			if (!polygon0->initialise())
-			{
-				ERR("Polygon::splitPolygonAtIntersection(): Initialization of polygon0 failed.");
-				exit (1);
-			}
+				polyline0.addPoint ((*polygon_it)->getPointID (k));
+			GeoLib::Polygon *polygon0(new Polygon(polyline0));
```
Comment (chleh) [inline GeoLib/Polygon.cpp]: I'd prefer unique_ptr here, and release if you transfer ownership.

File: GeoLib/Polygon.cpp [hunk 3]
```diff
@@ -368,43 +368,34 @@ void Polygon::splitPolygonAtIntersection (std::list<Polygon*>::iterator polygon_
...
-				polygon1->addPoint ((*polygon_it)->getPointID (k));
-			polygon1->addPoint (intersection_pnt_id);
-			if (!polygon1->initialise())
-			{
-				ERR("Polygon::splitPolygonAtIntersection(): Initialization of polygon1 failed.");
-				exit (1);
-			}
+				polyline1.addPoint ((*polygon_it)->getPointID (k));
+			polyline1.addPoint (intersection_pnt_id);
+			GeoLib::Polygon *polygon1(new Polygon(polyline1));
```
Comment (chleh) [inline GeoLib/Polygon.cpp]: Same ownership thing here.

File: GeoLib/Polygon.cpp (hunk 2 above)
Comment (TomFischer) [inline GeoLib/Polygon.cpp]: I found an other solution. See the comment in the commit e10ef00.

File: GeoLib/Polygon.cpp (hunk 3 above)
Comment (TomFischer) [inline GeoLib/Polygon.cpp]: Same solution as above.

File: GeoLib/AnalyticalGeometry.cpp (hunk 1 above)
Comment (TomFischer) [inline GeoLib/AnalyticalGeometry.cpp]: > Can v[0] be 0.0?

Yes, this could be the case. This was a serious problem. I hopefully fixed it in commit 6e75c9e?! Please have a look.

Comment (TomFischer) [inline GeoLib/AnalyticalGeometry.cpp]: > Btw: 0 <= qp[0] / v[0] essentially checks if qp[0] and v[0] have the same sign. Is that intended?

The intention here is to check if the quotient `qp[0] / v[0]` is in [0,1]. Maybe there is a better solution?

**Comment by TomFischer** [comment]
@rinkk I can not find your `while` loop comment anymore. Your point was valid. I removed the while loop in the commit e10ef00.

**Comment by chleh** [comment]
Your solution in e10ef00 looks valid. It looks much better than before, indeed. :+1:

**Comment by chleh** [comment]
If I see it correctly, you still divide by some vector components very often in 6e75c9e532cc4f50b3c0fbd784638b75ecfc21f2. And probably you always want to check if `0.0 <= a/b <= 1.0`.

> The intention here is to check if the quotient qp[0] / v[0] is in [0,1]. Maybe there is a better solution?

Let `a := qp[0]; b := v[0]`, then you could check:
(i) `0.0 <= a*b` That checks if a and b have the same sign and is logically the same as `0.0 <= a/b`, only it is guaranteed that you don't divide by 0.0. However, I don't know if that has any implications if `b == 0.0`, because then the new check would succeed whereas the old check would not, since `a/0.0 == NaN` which AFAIK compares `false` to everything.
(ii) `std::abs(a) <= std::abs(b)` is __always__ equivalent to `a/b <= 1.0`. There are no special considerations that must be taken for this case.

Let's discuss vis-a-vis.

**Comment by rinkk** [comment]
:+1:

File: GeoLib/Polygon.cpp [hunk 4]
```diff
@@ -348,64 +348,46 @@ void Polygon::ensureCWOrientation ()
...
+
+	GeoLib::Polyline polyline1{(*polygon_it)->getPointsVec()};
+	polyline1.addPoint(intersection_pnt_id);
+	for (std::size_t k(idx0 + 1); k <= idx1; k++)
+		polyline1.addPoint((*polygon_it)->getPointID(k));
+	polyline1.addPoint(intersection_pnt_id);
+
+	// remove original polyline and add two new polylines
+	std::list<GeoLib::Polygon*>::const_iterator polygon1_it =
+	    _simple_polygon_list.insert(_simple_polygon_list.erase(polygon_it),
```
Comment (endJunction) [inline GeoLib/Polygon.cpp]: Unfortunately the STL version shipped with gcc-4.8.1 does not implement the c++11 version of `std::list::erase` which is taking a const iterator. See http://en.cppreference.com/w/cpp/container/list/erase for details. This was only fixed with  gcc-4.9.0. Opening new issues to set minimum gcc version to 4.9.0.

Until the new gcc is required I would add an #if based on gcc version like this:
```c++
/// \todo Use const_iterator after requiring gcc >= 4.9.0
/// stl::erase takes const iterator required by c++11 since gcc-4.9.0
#if __GNUC__ <= 4 && (__GNUC_MINOR__ < 9)
			polygon1_it =
			    _simple_polygon_list.insert(_simple_polygon_list.erase(
			        (std::list<GeoLib::Polygon*>::iterator)polygon_it),
			        polygon1);
#else
			polygon1_it = _simple_polygon_list.insert(
			    _simple_polygon_list.erase(polygon_it), polygon1);
#endif
```

Comment (chleh) [inline GeoLib/Polygon.cpp]: Rather than introducing an `#if` here, I'd just go with a non-const iterator. I think using preprocessor macros here is worse than using non-const iterators. Remember: This is not an interface but an internal detail. Or is the build failing because of that?
Btw: you should probably use `const_cast<>`.

Comment (endJunction) [inline GeoLib/Polygon.cpp]: It's temporary solution until gcc 4.9 is used in CI environment.
Non-const iterator is visible in the interface, which is not good.
`const_cast<>` will probably not work because its type change from `std::list::iterator` to `std::list::const_iterator`

Comment (TomFischer) [inline GeoLib/Polygon.cpp]: Using the C style cast makes clang not happy.

Comment (endJunction) [inline GeoLib/Polygon.cpp]: You can make a new temporary std::list::iterator...

File: GeoLib/AnalyticalGeometry.cpp [hunk 5]
```diff
@@ -121,28 +121,55 @@ bool lineSegmentIntersect(
...
+		if (parallel(pq,v)) { // line segment (a,b) and (a,c) are also parallel
+			// Here it is already checked that the line segments (a,b) and (c,d)
+			// are parallel. At this point it is also known that the line
+			// segment (a,c) is also parallel to (a,b). In that case it is
+			// possible to express c as c(t) = a + t * (b-a) (analog for the
+			// point d). Since the evaluation of all three coordinate equations
+			// (x,y,z) have to lead to the same solution for the parameter t it
+			// is sufficient to evaluate t only once.
+
+			// Search id of coordinate with largest absolute value which is will
```
Comment (chleh) [inline GeoLib/AnalyticalGeometry.cpp]: ~~is~~

File: GeoLib/AnalyticalGeometry.cpp [hunk 6]
```diff
@@ -121,28 +121,55 @@ bool lineSegmentIntersect(
...
+			// be used in the subsequent computations. This prevents division by
+			// zero in case the line segments are parallel to one of the
+			// coordinate axis.
+			std::size_t i_max(std::abs(v[0]) <= std::abs(v[1]) ? 1 : 0);
+			i_max = std::abs(v[i_max]) <= std::abs(v[2]) ? 2 : i_max;
+			if (isLineSegmentIntersectingAB(qp, i_max)) {
+				s = c;
```
Comment (chleh) [inline GeoLib/AnalyticalGeometry.cpp]: Just wondering: Possibly ab and cd overlap in an entire range. Does the specific choice of s have any consequences? In principle you could use any point in `ab intersect cd`.

**Comment by chleh** [comment]
I don't know why, but I like your new intersection solution :laughing: 
:+1:

File: GeoLib/AnalyticalGeometry.cpp (hunk 6 above)
Comment (TomFischer) [inline GeoLib/AnalyticalGeometry.cpp]: In my opinion in this case the complete line segment `ab` and `cd` are intersecting each other should be returned. But this would be a separate PR.

**Comment by endJunction** [comment]
@chleh  This is much better solution using the `auto` then the proposed #if. Thanks!
